src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,320 @@
+                                const { 1: code } = match;
+                                // Update trial job's status based on result code
+                                if (parseInt(code, 10) === 0) {
+                                    environment.setStatus('SUCCEEDED');
+                                } else {
+                                    environment.setStatus('FAILED');
+                                }
+                                this.releaseEnvironmentResource(environment);
+                            }
+                        }
+                    }
+                } catch (error) {
+                    this.releaseEnvironmentResource(environment);
+                    this.log.error(`Update job status exception, error is ${error.message}`);
+                }
+            }
+        });
+    }
+
+    /**
+     * If a environment is finished, release the connection resource
+     * @param environment remote machine environment job detail
+     */
+    public releaseEnvironmentResource(environment: EnvironmentInformation): void {
+        const executorManager = this.environmentExecutorManagerMap.get(environment.id);
+        if (executorManager === undefined) {
+            throw new Error(`ExecutorManager is not assigned for environment ${environment.id}`);
+        }
+
+        // Note, it still keep reference in trialExecutorManagerMap, as there may be following requests from nni manager.
+        executorManager.releaseExecutor(environment.id);
+        const remoteEnvironment: RemoteMachineEnvironmentInformation = environment as RemoteMachineEnvironmentInformation;
+        if (remoteEnvironment.rmMachineMeta === undefined) {
+            throw new Error(`${remoteEnvironment.id} rmMachineMeta not initialized!`);
+        }
+        this.recycleMachineReservation(remoteEnvironment.rmMachineMeta);
+    }
+
+    public async startEnvironment(environment: EnvironmentInformation): Promise<void> {
+        if (this.sshConnectionPromises.length > 0) {
+            await Promise.all(this.sshConnectionPromises);
+            this.log.info('ssh connection initialized!');
+            // set sshConnectionPromises to [] to avoid log information duplicated
+            this.sshConnectionPromises = [];
+            if (this.trialConfig ===  undefined) {
+                throw new Error("trial config not initialized!");
+            }
+            Array.from(this.machineExecutorManagerMap.keys()).forEach(rmMeta => {
+                // initialize remoteMachineMetaOccupiedMap, false means not occupied
+                this.remoteMachineMetaOccupiedMap.set(rmMeta, false);
+            });
+        }
+        const remoteEnvironment: RemoteMachineEnvironmentInformation = environment as RemoteMachineEnvironmentInformation;
+        remoteEnvironment.status = 'WAITING';
+        await this.prepareEnvironment(remoteEnvironment);
+        await this.launchEnvironment(remoteEnvironment);
```

Comment:
It's called here, and inside of prepareEnvironment also. Is it expected? The logic can be clearer.

Reply:
refactored.